examples/cpp/pyperf/PyPerfNativeStackTrace.cc

```diff
@@ -42,26 +50,47 @@ NativeStackTrace::NativeStackTrace(uint32_t pid, const unsigned char *raw_stack,
   // The UPT implementation of these functions uses ptrace. We want to make sure they aren't getting called
   my_accessors.access_fpreg = NULL;
   my_accessors.resume = NULL;
-
+  unw_addr_space_t as;
+  unw_cursor_t cursor;
+  void *upt;
   int res;
-  unw_addr_space_t as = unw_create_addr_space(&my_accessors, 0);
-  void *upt = _UPT_create(pid);
-  if (!upt) {
-    this->symbols.push_back(std::string("[Error _UPT_create (system OOM)]"));
-    this->error_occurred = true;
-    goto out;
-  }
 
-  unw_cursor_t cursor;
-  // TODO: It's possible to make libunwind use cache using unw_set_caching_policy, which might lead to significent
-  //       performance improvement. We just need to make sure it's not dangerous. For now the overhead is good enough.
-  res = unw_init_remote(&cursor, as, upt);
-  if (res) {
-    std::ostringstream error;
-    error << "[Error unw_init_remote (" << unw_strerror(res) << ")]";
-    this->symbols.push_back(error.str());
-    this->error_occurred = true;
-    goto out;
+  // Pseudo-proactive way of implementing TTL - whenever any call is made, all expired entries are removed
```

Comment:
Together with TTL-based eviction - PyPerf already tracks PIDs exiting so we can (and should) make use of that - otherwise we risk PID reuse getting us false results.

See `populatePidTable` where in the `Pruning dead pids` step it removes PIDs. We should call `cache_delete_key` there as well. Please ensure it actually works in removing PIDs that exit

Reply:
I push a draft of implementation.
populatePIDTable calls static NativeStackTrace::Pruning_dead_pid(). Please take a look at the proposal and let me know if such an approach is acceptable. It is WIP, so please ignore debug logs. I will remove it later.

In the meantime, I am debugging why the cache version gives `missing` and do not manage to add native symbols properly.

Reply:
It's fine with me. Despite written in C++ this entire project is fairly functional.